src/renderer/atlas/AtlasEngine.r.cpp

```diff
@@ -0,0 +1,460 @@
+// The _api fields on the other hand are concurrently written to by others.
+
+#pragma warning(disable : 4100) // '...': unreferenced formal parameter
+// Disable a bunch of warnings which get in the way of writing performant code.
+#pragma warning(disable : 26429) // Symbol 'data' is never tested for nullness, it can be marked as not_null (f.23).
+#pragma warning(disable : 26446) // Prefer to use gsl::at() instead of unchecked subscript operator (bounds.4).
+#pragma warning(disable : 26481) // Don't use pointer arithmetic. Use span instead (bounds.1).
+#pragma warning(disable : 26482) // Only index into arrays using constant expressions (bounds.2).
+
+using namespace Microsoft::Console::Render;
+
+#pragma region IRenderEngine
+
+// Present() is called without the console buffer lock being held.
+// --> Put as much in here as possible.
+[[nodiscard]] HRESULT AtlasEngine::Present() noexcept
+try
+{
+    _adjustAtlasSize();
+    _reserveScratchpadSize(_r.maxEncounteredCellCount);
+    _processGlyphQueue();
+
+    if (WI_IsFlagSet(_r.invalidations, RenderInvalidations::Cursor))
+    {
+        _drawCursor();
+        WI_ClearFlag(_r.invalidations, RenderInvalidations::Cursor);
+    }
+
+    // The values the constant buffer depends on are potentially updated after BeginPaint().
+    if (WI_IsFlagSet(_r.invalidations, RenderInvalidations::ConstBuffer))
+    {
+        _updateConstantBuffer();
+        WI_ClearFlag(_r.invalidations, RenderInvalidations::ConstBuffer);
+    }
+
+    {
+#pragma warning(suppress : 26494) // Variable 'mapped' is uninitialized. Always initialize an object (type.5).
+        D3D11_MAPPED_SUBRESOURCE mapped;
+        THROW_IF_FAILED(_r.deviceContext->Map(_r.cellBuffer.get(), 0, D3D11_MAP_WRITE_DISCARD, 0, &mapped));
+        assert(mapped.RowPitch >= _r.cells.size() * sizeof(Cell));
+        memcpy(mapped.pData, _r.cells.data(), _r.cells.size() * sizeof(Cell));
+        _r.deviceContext->Unmap(_r.cellBuffer.get(), 0);
+    }
+
+    // After Present calls, the back buffer needs to explicitly be
+    // re-bound to the D3D11 immediate context before it can be used again.
+    _r.deviceContext->OMSetRenderTargets(1, _r.renderTargetView.addressof(), nullptr);
+    _r.deviceContext->Draw(3, 0);
+
+    // See documentation for IDXGISwapChain2::GetFrameLatencyWaitableObject method:
+    // > For every frame it renders, the app should wait on this handle before starting any rendering operations.
+    // > Note that this requirement includes the first frame the app renders with the swap chain.
+    assert(_r.frameLatencyWaitableObjectUsed);
+
+    // > IDXGISwapChain::Present: Partial Presentation (using a dirty rects or scroll) is not supported
+    // > for SwapChains created with DXGI_SWAP_EFFECT_DISCARD or DXGI_SWAP_EFFECT_FLIP_DISCARD.
+    // ---> No need to call IDXGISwapChain1::Present1.
+    //      TODO: Would IDXGISwapChain1::Present1 and its dirty rects have benefits for remote desktop?

```

Comment:
Solid maybe. Do note that the most recent versions of Remote Desktop have an automatic identification of areas of high motion on the screen and can effectively MPEG-video-streamify what's happening there to the remote client as a subsection of the display. I've successfully played Guild Wars 2 through an RDP session and watched YouTube videos through one at reasonable resolution (720p+). I bet we don't have to worry about this as much as we think we do (or did in the past).

Reply:
RDP maintainers confirmed poggers.